storage,kvserver: Improve SST collision checking for wide SSTs

itsbilal opened the pull request:
This change includes 4 changes to significantly improve
the performance of CheckSSTCollisions (and by extension,
AddSSTable) for cases where we add very wide sstables
relative to the engine:

1) Check if we're adding an sstable that has a small number of
keys, or has a 100x greater overlap with the engine relative
to its own size. In that case, switch to doing prefix seeks
inside CheckSSTCollisions, similar to what we did in #73514
and then reverted later).
2) Write bloom filters in L6 by default, to help speed up 1.
Also thread a new IterOption to only optionally read these
filter blocks in prefix iteration, defaulting to not reading
them. This iterator option hooks into the one added in
Pebble in cockroachdb/pebble#1685 ).
3) Add a ParseHook to allow command-line pebble options to
allow configuring the writing of bloom filters.
4) Update intentInterleavingIter to not prefix-seek the
intent/lock table iterator if we are in prefix seek mode and
the MVCC iterator returned nothing.

Fixes #80980.

Release note (performance improvement): Significantly improve
performance of IMPORTs when the source is producing data
not sorted by the destination table's primary key, especially
if the destination table has a very large primary key with
lots of columns.

Commit:
storage,kvserver: Improve SST collision checking for wide SSTs

This change includes 4 changes to significantly improve
the performance of CheckSSTCollisions (and by extension,
AddSSTable) for cases where we add very wide sstables
relative to the engine:

1) Check if we're adding an sstable that has a small number of
keys, or has a 100x greater overlap with the engine relative
to its own size. In that case, switch to doing prefix seeks
inside CheckSSTCollisions, similar to what we did in #73514
and then reverted later).
2) Thread a new IterOption to only optionally read L6
filter blocks in prefix iteration, defaulting to not reading
them. This iterator option hooks into the one added in
Pebble in cockroachdb/pebble#1685 ).
3) Add a ParseHook to allow command-line pebble options to
allow configuring the writing of bloom filters. This allows
L6 filter block writing to be turned on through a command
line argument.
4) Update intentInterleavingIter to not prefix-seek the
intent/lock table iterator if we are in prefix seek mode and
the MVCC iterator returned nothing.

Fixes #80980.

Release note (performance improvement): Significantly improve
performance of IMPORTs when the source is producing data
not sorted by the destination table's primary key, especially
if the destination table has a very large primary key with
lots of columns.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81062)
<!-- Reviewable:end -->

itsbilal commented:
Actually, it'd be better for me to split this into multiple commits. At least the write-L6-bloom-filters-by-default piece can be moved into its own commit, as that'll be the only thing not backported to 22.1.

sumeerbhola reviewed (changes requested):
> Reviewed 13 of 13 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NhTis2h6pxfv7PFGy:-N2NhTit3p3brgEBCZt3:bwtl2ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L172)):*
> ```Go
> 		bytes, err := cArgs.EvalCtx.GetApproximateDiskBytes(start.Key, end.Key)
> 		if err == nil {
> 			usePrefixSeek = bytes > 100*uint64(len(sst))
> ```

Can you add some code commentary about how these two constants were chosen. It will help someone in the future if they want to refine the heuristic.
___
*[`pkg/server/config.go` line 636 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NiDMO91dq2NpvJ4NF:-N2NiDMO91dq2NpvJ4NG:bkk5t9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/server/config.go#L636)):*
> ```Go
> 						case "none":
> 							return nil, nil
> 						case "rocksdb.BuiltinBloomFilter":
> ```

we already do this in `DefaultPebbleOptions`, so I am confused on why we are adding this.
___
*[`pkg/storage/intent_interleaving_iter.go` line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NjA_E0QHVg_g4Pumw:-N2NjA_E0QHVg_g4Pumx:b9uwpx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/intent_interleaving_iter.go#L331)):*
> ```Go
> 		intentSeekKey = nil
> 		i.intentKey = nil
> 	}
> ```

are we covering this in the datadriven test?
___
*[`pkg/storage/pebble.go` line 637 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nmt66BmkOvmLkWWtG:-N2Nmt66BmkOvmLkWWtH:b-74vnmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a30f5a33ce496358edf3928b5f029a87519335d4/pkg/storage/pebble.go#L637)):*
> ```Go
> 	// of the benefit of having bloom filters on every level for only 10% of the
> 	// memory cost.
> 	opts.Levels[6].FilterPolicy = nil
> ```

don't we need a way for someone to configure that they want L0-L5 bloomfilters, but not L6 bloomfilters?
___
*[`pkg/storage/sst.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NlfsLC0RIgm5_LzQI:-N2NlfsLC0RIgm5_LzQJ:bf2yk9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L74)):*
> ```Go
> 	}
> 
> 	extIter := reader.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
> ```

nit: can this extIter initialization be moved down after the sstIter initialization, so it is all in one place?
___
*[`pkg/storage/sst.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmEoM5jI9juIa9ouU:-N2NmEoM5jI9juIa9ouV:bnn81rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L102)):*
> ```Go
> 		extOK, extErr = extIter.Valid()
> 	}
> 	for sstErr == nil && sstOK && (usePrefixSeek || (extOK && extErr == nil)) {
> ```

Can you try coding separate loops for the two cases and see if it looks simple and doesn't have too much code duplication. The loop here is confusing me -- large parts are gated by `usePrefixSeek` and the part that does the coordinated seeks isn't gated by `!seekPrefixGE` which is also confusing.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go` line 703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk6FI4jLOppXRX9zi:-N2Nk6FI4jLOppXRX9zj:b-hwljf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L703)):*
> ```Go
> 						// non-prefix Seek in the conflict check. Sending in nil stats
> 						// makes this easier as that forces the function to rely exclusively
> 						// on ApproxDiskBytes.
> ```

I didn't understand this comment. I thought `approxDiskBytes==0` is when we want the SeekPrefixGE to be used. And why don't we always compute the MVCCStats for the sst?
___
*[`pkg/storage/sst_test.go` line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NjKG90DUU9T1ASVFr:-N2NjKG90DUU9T1ASVFs:b3edv1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst_test.go#L90)):*
> ```Go
> 					startKey, endKey := MVCCKey{Key: roachpb.Key(start)}, MVCCKey{Key: roachpb.Key(end)}
> 					_, err := CheckSSTConflicts(ctx, sstFile.Bytes(), engine, startKey, endKey,
> 						false /*disallowShadowing*/, hlc.Timestamp{} /*disallowShadowingBelow*/, tc.maxIntents, false /* usePrefixSeek */)
> ```

are we not testing the true case?


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> Reviewed 6 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal)

___
*[`pkg/storage/engine.go` line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk_EI3Ga0dE9dbFca:-N2Nk_EI3Ga0dE9dbFcb:buk426x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/engine.go#L295)):*
> ```Go
> 	// is minimized if the probability of the key existing is not low or if
> 	// this is a one-time Seek (where loading the data block directly is better).
> 	UseL6Filters bool
> ```

Is this expected to be used outside of `storage.CheckSSTConflicts`? I think we should generally err towards fewer options and knobs exposed through the storage package interface. Can we unexport it?
___
*[`pkg/storage/sst.go` line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmcocBupt9jMfX0d2:-N2NmcocBupt9jMfX0d3:bm5es3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L110)):*
> ```Go
> 			// key in sstIter. Note that we can't just use an exhausted extIter
> 			// as a sign that we are done; extIter could be skipping keys, so it
> 			// must be re-seeked.
> ```

by "skipping keys" this comment is referring to the fact that exhaustion of a prefix iterator indicates the prefix is exhausted, not that there are no keys after the prefix? I think it could use some rephrasing.


<!-- Sent from Reviewable.io -->

nicktrav reviewed (commented):
> Reviewed 13 of 13 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`-- commits` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NuQZu7_c6NwYS_OXA):*
> mitigated

nit: is it worth mentioning the tradeoff of additional disk usage due to the writing of these L6 filter blocks?

Also - is this worth of a release note, given it is "visible" under the interpretation that an operator would observe a material disk usage increase after this change.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NhTis2h6pxfv7PFGy:-N2NqmH75f2ZY_3Vymvp:b-vcxmqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L172)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add some code commentary about how these two constants were chosen. It will help someone in the future if they want to refine the heuristic.
</blockquote></details>

In addition to the commentary, mind lifting to a `const` (either in this method, or outside of it). Will focus the commentary a bit more.
___
*[`pkg/storage/sst.go` line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Ns3oV3B78RG0ecYVI:-N2Ns3oV3B78RG0ecYVJ:bq117d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L67)):*
> ```Go
> 		nonPrefixIter := reader.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{UpperBound: end.Key})
> 		nonPrefixIter.SeekGE(start)
> 		valid, _ := nonPrefixIter.Valid()
> ```

nit: store err here and return a few lines below? IIUC `valid` will be false if `err != nil`, so you don't need to check the err.


<!-- Sent from Reviewable.io -->

Commit:
storage: Enable writing of L6 filter blocks by default.

Previously, we did not write bloom filters on L6 sstables.
The main worry here was the higher memory usage and the
overhead of loading a large filter block for the large L6
SSTs in memory. This is now mitigated thanks to the optional
reading of L6 filter blocks only in narrow use cases where
they're most helpful.

Release note: None.

itsbilal reviewed (commented):
TFTRs!

I actually hadn't run the microbenchmark earlier for figuring out a good threshold to set `usePrefixSeek`. Here it is. I just adapted `BenchmarkCheckSSTCollisions` to test along the overlap bytes axis, not along the earlier number-of-versions axis. Old = usePrefixSeek false, new = usePrefixSeek true:

```
CheckSSTConflicts/keys=1000/sstKeys=10/overlap=true/prefixSeek=false-24         76.6µs ± 1%   69.7µs ± 2%    -9.01%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=1000/sstKeys=200/overlap=true/prefixSeek=false-24         379µs ± 1%    181µs ± 1%   -52.07%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=1000/sstKeys=1000/overlap=true/prefixSeek=false-24        887µs ± 2%    654µs ± 1%   -26.23%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=1000/sstKeys=10000/overlap=true/prefixSeek=false-24      2.13ms ± 1%   5.90ms ± 1%  +177.30%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=10000/sstKeys=10/overlap=true/prefixSeek=false-24        76.7µs ± 1%   67.9µs ± 1%   -11.44%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=10000/sstKeys=200/overlap=true/prefixSeek=false-24        414µs ± 1%    181µs ± 1%   -56.20%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=10000/sstKeys=1000/overlap=true/prefixSeek=false-24      1.76ms ± 1%   0.66ms ± 1%   -62.53%  (p=0.000 n=9+10)
CheckSSTConflicts/keys=10000/sstKeys=10000/overlap=true/prefixSeek=false-24     7.96ms ± 2%   5.99ms ± 1%   -24.70%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=100000/sstKeys=10/overlap=true/prefixSeek=false-24       73.8µs ± 1%   66.3µs ± 1%   -10.16%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=100000/sstKeys=200/overlap=true/prefixSeek=false-24       472µs ± 0%    176µs ± 0%   -62.63%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=100000/sstKeys=1000/overlap=true/prefixSeek=false-24     2.09ms ± 1%   0.65ms ± 0%   -68.81%  (p=0.000 n=10+8)
CheckSSTConflicts/keys=100000/sstKeys=10000/overlap=true/prefixSeek=false-24    19.4ms ± 1%    6.1ms ± 1%   -68.84%  (p=0.000 n=10+9)
CheckSSTConflicts/keys=1000/sstKeys=100/overlap=true/prefixSeek=false-24         219µs ± 0%    123µs ± 1%   -44.12%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=1000/sstKeys=100000/overlap=true/prefixSeek=false-24     6.97ms ± 0%  58.27ms ± 0%  +735.83%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=10000/sstKeys=100/overlap=true/prefixSeek=false-24        240µs ± 2%    123µs ± 1%   -48.57%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=10000/sstKeys=100000/overlap=true/prefixSeek=false-24    21.5ms ± 1%   59.2ms ± 2%  +174.97%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=100000/sstKeys=100/overlap=true/prefixSeek=false-24       257µs ± 2%    120µs ± 2%   -53.25%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=100000/sstKeys=100000/overlap=true/prefixSeek=false-24   81.3ms ± 1%   60.5ms ± 0%   -25.59%  (p=0.008 n=5+5)
```

I've split it up between overlap ratios between sst keys (numerator) and engine keys (denominator):

```
1/10000:
CheckSSTConflicts/keys=100000/sstKeys=10/overlap=true/prefixSeek=false-24       73.8µs ± 1%   66.3µs ± 1%   -10.16%  (p=0.000 n=10+10)

1/1000:
CheckSSTConflicts/keys=10000/sstKeys=10/overlap=true/prefixSeek=false-24        76.7µs ± 1%   67.9µs ± 1%   -11.44%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=100000/sstKeys=100/overlap=true/prefixSeek=false-24       257µs ± 2%    120µs ± 2%   -53.25%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=100000/sstKeys=200/overlap=true/prefixSeek=false-24       472µs ± 0%    176µs ± 0%   -62.63%  (p=0.008 n=5+5)

1/100:
CheckSSTConflicts/keys=1000/sstKeys=10/overlap=true/prefixSeek=false-24         76.6µs ± 1%   69.7µs ± 2%    -9.01%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=10000/sstKeys=100/overlap=true/prefixSeek=false-24        240µs ± 2%    123µs ± 1%   -48.57%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=10000/sstKeys=200/overlap=true/prefixSeek=false-24        414µs ± 1%    181µs ± 1%   -56.20%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=100000/sstKeys=1000/overlap=true/prefixSeek=false-24     2.09ms ± 1%   0.65ms ± 0%   -68.81%  (p=0.000 n=10+8)

1/10:
CheckSSTConflicts/keys=1000/sstKeys=100/overlap=true/prefixSeek=false-24         219µs ± 0%    123µs ± 1%   -44.12%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=1000/sstKeys=200/overlap=true/prefixSeek=false-24         379µs ± 1%    181µs ± 1%   -52.07%  (p=0.008 n=5+5)
CheckSSTConflicts/keys=10000/sstKeys=1000/overlap=true/prefixSeek=false-24      1.76ms ± 1%   0.66ms ± 1%   -62.53%  (p=0.000 n=9+10)
CheckSSTConflicts/keys=100000/sstKeys=10000/overlap=true/prefixSeek=false-24    19.4ms ± 1%    6.1ms ± 1%   -68.84%  (p=0.000 n=10+9)

1/1:
CheckSSTConflicts/keys=1000/sstKeys=1000/overlap=true/prefixSeek=false-24        887µs ± 2%    654µs ± 1%   -26.23%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=10000/sstKeys=10000/overlap=true/prefixSeek=false-24     7.96ms ± 2%   5.99ms ± 1%   -24.70%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=100000/sstKeys=100000/overlap=true/prefixSeek=false-24   81.3ms ± 1%   60.5ms ± 0%   -25.59%  (p=0.008 n=5+5)

10/1:
CheckSSTConflicts/keys=1000/sstKeys=10000/overlap=true/prefixSeek=false-24      2.13ms ± 1%   5.90ms ± 1%  +177.30%  (p=0.000 n=10+10)
CheckSSTConflicts/keys=10000/sstKeys=100000/overlap=true/prefixSeek=false-24    21.5ms ± 1%   59.2ms ± 2%  +174.97%  (p=0.008 n=5+5)

100/1:
CheckSSTConflicts/keys=1000/sstKeys=100000/overlap=true/prefixSeek=false-24     6.97ms ± 0%  58.27ms ± 0%  +735.83%  (p=0.008 n=5+5)
```

Excluding the tiny-sst cases (where very little time is spent iterating through keys), we see that cases where the sst:engine overlap is 1:10 or lower (i.e. more engine keys per sst key), then there's a clear >50% benefit to using prefix seeks. At 1:1 it's more of a wash, and when the ratio is more sst-heavy, it's definitely better to do the regular seek.

Maybe this calls for setting the ratio to 10 instead of the currently-conservative 100. I could rerun my end-to-end experiments with the const at 10, but I'm sure it won't change anything because all the overlap ratios (in bytes) that I saw there were in the 200-10,000 range.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-:-N2Xkn3e4lqF687y-d4_:b-kxp24g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, @nicktrav, and @sumeerbhola)

___
*[`-- commits` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NuQZu7_c6NwYS_OXA):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

> mitigated

nit: is it worth mentioning the tradeoff of additional disk usage due to the writing of these L6 filter blocks?

Also - is this worth of a release note, given it is "visible" under the interpretation that an operator would observe a material disk usage increase after this change.
</blockquote></details>

I don't think that'd be necessary. We saw only a 1% increase in space amplification in L6 with filter blocks enabled. For the LSm as a whole, the % increase will be even lower. The memory usage of repeatedly reading that block is a bigger concern and that's why we mitigated it with that IterOption.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NhTis2h6pxfv7PFGy:-N2Xklm2BYY38CayCgG5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L172)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

In addition to the commentary, mind lifting to a `const` (either in this method, or outside of it). Will focus the commentary a bit more.
</blockquote></details>

Done.
___
*[`pkg/server/config.go` line 636 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NiDMO91dq2NpvJ4NF:-N2NqRdj2O0qeEocz9By:b-uij32r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/server/config.go#L636)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we already do this in `DefaultPebbleOptions`, so I am confused on why we are adding this.
</blockquote></details>

The goal is to be able to opt into L6 bloom filters with a command line argument. Currently you can't do that because our parser ignores filter policies. A customer that wants to do a big import on 22.1 could opt into it by passing this flag, without us changing the default behaviour.

The `DefaultPebbleOptions` change is in the second commit which won't be backported. 
___
*[`pkg/storage/engine.go` line 295 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk_EI3Ga0dE9dbFca:-N2Nqusw8H_rShljvvAb:b-hjs9fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/engine.go#L295)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Is this expected to be used outside of `storage.CheckSSTConflicts`? I think we should generally err towards fewer options and knobs exposed through the storage package interface. Can we unexport it?
</blockquote></details>

Unexported. Feels a little weird to have an unexported field here but it's better than the alternative.
___
*[`pkg/storage/intent_interleaving_iter.go` line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NjA_E0QHVg_g4Pumw:-N2Y4-qaAHoal5xUxgN6:bk1hleh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/intent_interleaving_iter.go#L331)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are we covering this in the datadriven test?
</blockquote></details>

yep, `pkg/storage/testdata/intent_interleaving_iter/basic:143` tests this already.
___
*[`pkg/storage/pebble.go` line 637 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nmt66BmkOvmLkWWtG:-N2NrB6i2HeR62s6HXAb:bs4a69a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a30f5a33ce496358edf3928b5f029a87519335d4/pkg/storage/pebble.go#L637)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

don't we need a way for someone to configure that they want L0-L5 bloomfilters, but not L6 bloomfilters?
</blockquote></details>

They can do that by passing in a `[Level 6]TableFilter=none` pebble option. But we're changing the default write-time behaviour on master (after 22.1), as the disk space amplification is minimal (1%). The bigger worry is read-time memory usage, which we'll mitigate with the IterOption defaulting to false except in `CheckSSTCollisions`.
___
*[`pkg/storage/sst.go` line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Ns3oV3B78RG0ecYVI:-N2XsK3_54yM9wmxJJ_s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L67)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: store err here and return a few lines below? IIUC `valid` will be false if `err != nil`, so you don't need to check the err.
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NlfsLC0RIgm5_LzQI:-N2Y5xnjC-DTk1eOZhg5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L74)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: can this extIter initialization be moved down after the sstIter initialization, so it is all in one place?
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmEoM5jI9juIa9ouU:-N2Y7_G869RfyrmK5F7e:bivsm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L102)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you try coding separate loops for the two cases and see if it looks simple and doesn't have too much code duplication. The loop here is confusing me -- large parts are gated by `usePrefixSeek` and the part that does the coordinated seeks isn't gated by `!seekPrefixGE` which is also confusing.
</blockquote></details>

There are only two places in the loop where we use `if usePrefixSeek`, so the separate loop thing duplicates a _lot_ of code. I did add more comments if that helps. What part are you referring to where we do the coordinated seeks? The one where `cmp < 0` down below? In that case we would be seeking the engine iterator forward either way. I could maybe rewrite that if statement, but you'd still need a conditional for the `sstIter.NextKey()` part

Any of the cases near the bottom where we're calling a Next on the sst iter and then seeking the engine iterator, we don't need to do anything as those cases will work as-is for both kinds of seeks. The `trySeekUsingNext` optimization also helps a lot there.
___
*[`pkg/storage/sst.go` line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmcocBupt9jMfX0d2:-N2Y4aqd84VdH1fFNU1q:b-hqpg6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L110)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

by "skipping keys" this comment is referring to the fact that exhaustion of a prefix iterator indicates the prefix is exhausted, not that there are no keys after the prefix? I think it could use some rephrasing.
</blockquote></details>

Yep I meant exhaustion mostly, but also I just generally meant that we can't rely on prefix iteration to give us all keys. Clarifying.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go` line 703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk6FI4jLOppXRX9zi:-N2XsPys7x2Jtf5quuFv:b-2c4ryc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L703)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I didn't understand this comment. I thought `approxDiskBytes==0` is when we want the SeekPrefixGE to be used. And why don't we always compute the MVCCStats for the sst?
</blockquote></details>

approxDiskBytes=0 is no overlap with the engine, so we want a regular seek to just skip past everything.

Not sending in stats is a weird test-only knob to get around the "if less than 100 keys then always prefix seek" check in cmd_add_sstable, because that relies on the stats to be passed in. We still calculate stats but further down in that function. That's what the comment is talking about. I'll clarify it a little bit.
___
*[`pkg/storage/sst_test.go` line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NjKG90DUU9T1ASVFr:-N2NqHyq4kito2tREV_t:bxjrnns) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst_test.go#L90)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are we not testing the true case?
</blockquote></details>

We were, in the EvalAddSSTable test. But now I fixed it here too, good catch.


<!-- Sent from Reviewable.io -->

nicktrav reviewed (approved):
All my feedback addressed. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 12 of 12 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-:-N2YEYYP6DN_b79ixrNM:baokucm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2YDHX3CmDAM0QnIe37:-N2YDHX4-EylYsHzKQk7:b-8jg688) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcdf7d0d0922f317e2abfd8cc5971db60d8d792d/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L89)):*
> ```Go
> // i.e. the engine:sst byte/key ratio is high and skewed in favour of the engine.
> // In those cases, since we are less likely to skip SST keys with regular seeks
> // on the engine, it is more efficient to utilize bloom filters the engine's SSTs
> ```

nit: "utilize bloom filters _on_ the engine's SSTs"?


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @sumeerbhola)

___
*[`pkg/storage/pebble_iterator.go` line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2luUAS3U9GmECOpRyP:-N2luUAS3U9GmECOpRyQ:ba7qz53) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a9a92a2253affb4d594c320a1b915785d5056c/pkg/storage/pebble_iterator.go#L129)):*
> ```Go
> 	// through in pebble.
> 	//
> 	// p.options.useL6Filters = opts.useL6Filters
> ```

Pebble bump is merged now. :+1:
___
*[`pkg/storage/sst.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmEoM5jI9juIa9ouU:-N2lum7mEn3Ad-CngKsv:b-ic2xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L102)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

There are only two places in the loop where we use `if usePrefixSeek`, so the separate loop thing duplicates a _lot_ of code. I did add more comments if that helps. What part are you referring to where we do the coordinated seeks? The one where `cmp < 0` down below? In that case we would be seeking the engine iterator forward either way. I could maybe rewrite that if statement, but you'd still need a conditional for the `sstIter.NextKey()` part

Any of the cases near the bottom where we're calling a Next on the sst iter and then seeking the engine iterator, we don't need to do anything as those cases will work as-is for both kinds of seeks. The `trySeekUsingNext` optimization also helps a lot there.
</blockquote></details>

Since the `usePrefixSeek`-gated sections are concentrated at the beginning, where we're positioning the iterators, maybe there's a split implementation that factors out the remainder of the loop into a `compareForConflicts` function?
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go` line 703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk6FI4jLOppXRX9zi:-N2lvd2S9ZKtTguy_3rb:b-g14hhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L703)):*
> approxDiskBytes=0 is no overlap with the engine, so we want a regular seek to just skip past everything.

Do we know how frequent this case is? Is it only relevant for AddSSTables appending to the very end of the keyspace? For the same reason snapshots tend to end up in L0 (#80589), I'd expect boundary overlap would produce small non-zero byte estimates for key ranges other than those at the very end of the keyspace (eg, during an ordered import into a single table with no indexes).


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-:-N2qS-2v2OdvYiDkCjIp:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2YDHX3CmDAM0QnIe37:-N2qS0Nd0mk8xrxMVYEZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcdf7d0d0922f317e2abfd8cc5971db60d8d792d/pkg/kv/kvserver/batcheval/cmd_add_sstable.go#L89)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: "utilize bloom filters _on_ the engine's SSTs"?
</blockquote></details>

Done.
___
*[`pkg/storage/pebble_iterator.go` line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2luUAS3U9GmECOpRyP:-N2qS7gbCn-Lof0jYuKZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40a9a92a2253affb4d594c320a1b915785d5056c/pkg/storage/pebble_iterator.go#L129)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Pebble bump is merged now. :+1:
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2NmEoM5jI9juIa9ouU:-N2qqDm0A521OALgiTlK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/storage/sst.go#L102)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Since the `usePrefixSeek`-gated sections are concentrated at the beginning, where we're positioning the iterators, maybe there's a split implementation that factors out the remainder of the loop into a `compareForConflicts` function?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go` line 703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2Nk6FI4jLOppXRX9zi:-N2qSBG90a48ANvwm_sD:b8m8c00) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a3240f11d52dcdc973785fc080b71b37c86304c/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L703)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

> approxDiskBytes=0 is no overlap with the engine, so we want a regular seek to just skip past everything.

Do we know how frequent this case is? Is it only relevant for AddSSTables appending to the very end of the keyspace? For the same reason snapshots tend to end up in L0 (#80589), I'd expect boundary overlap would produce small non-zero byte estimates for key ranges other than those at the very end of the keyspace (eg, during an ordered import into a single table with no indexes).
</blockquote></details>

I saw it a lot for sorted imports, where the overlap was straight up 0 in the engine. More commonly than when the overlap was a small nonzero value. Either way this is just a special case in the test, the actual code in practice doesn't special-case == 0.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 4 of 12 files at r3, 3 of 6 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/sst.go` line 86 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDDXPBq3zaCIfQq2f:-N2wDDXQ3XoVZ2nm1t7q:b-fkk3ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L86)):*
> ```Go
> 	}
> 	defer sstIter.Close()
> 	sstIter.SeekGE(start)
> ```

can you move this `sstIter.SeekGE` to after `compareForCollision` so it is in the same place as `sstIter.Valid()`?
___
*[`pkg/storage/sst.go` line 224 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDjnTAjEgA14-RMWA:-N2wDjnTAjEgA14-RMWB:bqwvsku) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L224)):*
> ```Go
> 			if err := ctx.Err(); err != nil {
> 				return enginepb.MVCCStats{}, err
> 			}
> ```

Can we move the following here and not do it before the for loop?
```
extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
extOK, extErr = extIter.Valid()
```
___
*[`pkg/storage/sst.go` line 234 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDzSsC8PLxp8TGV1E:-N2wDzSsC8PLxp8TGV1F:b-67otyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L234)):*
> ```Go
> 				sstOK, sstErr = sstIter.Valid()
> 				if sstOK {
> 					extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
> ```

this seek could happen in the next iteration instead of being coded here.

Also, we seem to be ignoring the case where `extErr != nil`.
___
*[`pkg/storage/sst.go` line 244 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wEnNA-aZChPiSsaJl:-N2wEnNA-aZChPiSsaJm:bn7zlwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L244)):*
> ```Go
> 
> 			// Keep seeking the iterators until both keys are equal.
> 			if cmp := bytes.Compare(extKey.Key, sstKey.Key); cmp < 0 {
> ```

aren't the cases `cmp < 0` and `cmp > 0` an internal error since we are doing prefix iteration and `SeekPrefixGE` guarantees that the prefix must match?


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-:-N2wWeAn5ysW9YXf5o-E:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/sst.go` line 86 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDDXPBq3zaCIfQq2f:-N2wWc0j98NHBncoVABr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L86)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you move this `sstIter.SeekGE` to after `compareForCollision` so it is in the same place as `sstIter.Valid()`?
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 224 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDjnTAjEgA14-RMWA:-N2wWctTBo_qia33Hsk4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L224)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can we move the following here and not do it before the for loop?
```
extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
extOK, extErr = extIter.Valid()
```
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 234 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wDzSsC8PLxp8TGV1E:-N2wWbKTAKcazNrNmCUr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L234)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this seek could happen in the next iteration instead of being coded here.

Also, we seem to be ignoring the case where `extErr != nil`.
</blockquote></details>

Done.
___
*[`pkg/storage/sst.go` line 244 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wEnNA-aZChPiSsaJl:-N2wWGR2AlMAH3IG0LuH:bbs4b76) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L244)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

aren't the cases `cmp < 0` and `cmp > 0` an internal error since we are doing prefix iteration and `SeekPrefixGE` guarantees that the prefix must match?
</blockquote></details>

Not necessarily, since the prefix is defined as only the user key part of the MVCC key and we could still have cmp < 0 or cmp > 0 for cases where we find a different version of the other key.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/sst.go` line 244 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wEnNA-aZChPiSsaJl:-N2x3p692X1aXehx3xt6:b-t2kmc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L244)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Not necessarily, since the prefix is defined as only the user key part of the MVCC key and we could still have cmp < 0 or cmp > 0 for cases where we find a different version of the other key.
</blockquote></details>

But `bytes.Compare(extKey.Key, sstKey.Key)` is comparing the `roachpb.Key`, which is the prefix.
___
*[`pkg/storage/sst.go` line 212 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2x2sieBW1B_WxypZ-P:-N2x2sieBW1B_WxypZ-Q:bmbmgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc37cce8c41662c2765fae27bb3846fde5050871/pkg/storage/sst.go#L212)):*
> ```Go
> 		if sstOK {
> 			extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
> 		}
> ```

I was suggesting removing this SeekGE since we now do it at the beginning of the for loop.
___
*[`pkg/storage/sst.go` line 213 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2x3BbzALo5jZp6kJU5:-N2x3BbzALo5jZp6kJU6:b-kjieae) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc37cce8c41662c2765fae27bb3846fde5050871/pkg/storage/sst.go#L213)):*
> ```Go
> 			extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
> 		}
> 		// In the case of prefix seeks, only look at sst iterator exhaustion/errors.
> ```

... seeks, do not look at engine prefix iter exhaustion. This is because ...

(since we do want to look at engine prefix errors)


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/sst.go` line 244 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2wEnNA-aZChPiSsaJl:-N2xdZys2X-8I7UI5KBG:b35wj2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbcf2e04a3cda9e129385cf238ba1085195eb147/pkg/storage/sst.go#L244)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

But `bytes.Compare(extKey.Key, sstKey.Key)` is comparing the `roachpb.Key`, which is the prefix.
</blockquote></details>

Done. Good point.
___
*[`pkg/storage/sst.go` line 212 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2x2sieBW1B_WxypZ-P:-N2xdYEGDXG2xq8A2_-F:biup46w) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc37cce8c41662c2765fae27bb3846fde5050871/pkg/storage/sst.go#L212)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I was suggesting removing this SeekGE since we now do it at the beginning of the for loop.
</blockquote></details>

Oops, yeah my bad. Done.
___
*[`pkg/storage/sst.go` line 213 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2x3BbzALo5jZp6kJU5:-N2xdZQY8eFxBmEtfFIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc37cce8c41662c2765fae27bb3846fde5050871/pkg/storage/sst.go#L213)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

... seeks, do not look at engine prefix iter exhaustion. This is because ...

(since we do want to look at engine prefix errors)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 12 files at r3, 1 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-:-N2xgiWO-XQKJnLui2hF:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @jbowens, @nicktrav, and @sumeerbhola)

___
*[`pkg/storage/sst.go` line 242 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81062#-N2xgUBp4n4M7wyNXICM:-N2xgUBp4n4M7wyNXICN:bmul18x) ([raw file](https://github.com/cockroachdb/cockroach/blob/555e92cfc04dae2d6cb32f3085ca39ab36c659bc/pkg/storage/sst.go#L242)):*
> ```Go
> 			// We just seeked the engine iter. If it has a mismatching prefix,
> 			// the iterator is not obeying its contract.
> 			if cmp := bytes.Compare(extKey.Key, sstKey.Key); cmp != 0 {
> ```

nit: could use `bytes.Equal`.


<!-- Sent from Reviewable.io -->

itsbilal commented:
TFTR!

bors r=sumeerbhola,nicktrav
